Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,91 @@
 
 #include "config.h"
 #include "ScrollAnchoringController.h"
+#include "ElementIterator.h"
+#include "HTMLHtmlElement.h"
+#include "RenderBox.h"
+#include "TypedElementDescendantIteratorInlines.h"
+#include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
 void ScrollAnchoringController::invalidateAnchorElement()
 {
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    if (!m_midUpdatingScrollPositionForAnchorElement) {
+        m_anchorElement = nullptr;
+        m_lastOffsetForAnchorElement = { };
+        m_isQueuedForScrollPositionUpdate = false;
+        m_frameView.queueScrollableAreaForScrollAnchoringUpdate(m_frameView);
+    }
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
 {
-    // TODO: implement
-    // perform Anchor Element Selection algorithm on dom elements within frame view's viewable region
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return ScrollOffset(FloatPoint(candidate.absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location()));
```

Comment:
Is this right with pinch and command-+ zooming?

Reply:
Is it also really a ScrollOffset (i.e. always positive) in RTL documents?

Reply:
This bit will need to be refactored in https://github.com/WebKit/WebKit/pull/17934 so we can investigate this a bit more once we flesh out the coordinate system we are going to use for subscrollers.